core/src/main/java/io/grpc/internal/JsonUtil.java

```diff
@@ -96,57 +96,92 @@
 
   /**
    * Gets a number from an object for the given key.  If the key is not present, this returns null.
-   * If the value is not a Double, throws an exception.
+   * If the value does not represent a double, throws an exception.
    */
   @Nullable
-  public static Double getNumber(Map<String, ?> obj, String key) {
+  public static Double getNumberAsDouble(Map<String, ?> obj, String key) {
```

Comment:
1) Convenience. We need clear error messages when a field is not present. Getting a field from an object is the strongly-common case, so it would be worth the duplicate method
2) The only other way to get a number is from a list, and we tend to have specialized helpers for that (`getListOfObects`, `getListOfStrings`). It is rare (currently non-existent) to need to mix-and-match types within a list.

Reply:
Maybe something like this?

```java
  @Nullable
  public static Double getNumberAsDoubleFromObj(Map<String, ?> obj, String key) {
    try {
      return getNumberAsDouble(getFromObj(obj, key));
    } catch (NumberFormatException e) {
      throw new IllegalArgumentException(String.format("key '%s': %s", key, e.getMessage()));
    } catch (IllegalArgumentException e) {
      throw new IllegalArgumentException(
          String.format("key '%s' in '%s':  %s", key, obj, e.getMessage()));
    }
  }

  public static Double getNumberAsDouble(Object value) {
    if (value instanceof Double) {
      return (Double) value;
    }
    if (value instanceof String) {
      try {
        return Double.parseDouble((String) value);
      } catch (NumberFormatException e) {
        throw new NumberFormatException(
            String.format("value '%s' is not a string representation of a double", value));
      }
    }
    throw new IllegalArgumentException(String.format("value '%s' is not a number", value));
  }

 static Object getFromObj(Map<String, ?> obj, String key) {
    assert key != null;
    if (!obj.containsKey(key)) {
      return null;
    }
    return obj.get(key);
  }
```

Pros:
1) we get separation of concerns
2) `getFromObj` (name tentative) can be reused in all `getNumberAs*FromObj`
3) `getNumberAs*` can be reused elsewhere
4) `getNumberAs*` are not nullable
5) straightforward tests

Reply:
I don't understand what `getFromObj()` is providing, as `obj.get(key)` return `null` for non-present keys. This isn't Python. Note, if we had such a method it should throw an exception instead of returning null. The current methods consider null to be just another type and so fall-through to catch-all exceptions.

Nobody should use `getNumberAsDouble(Object)`; "you're doing it wrong" and the caller will probably have poor error handling. I'm going to argue pretty strongly that we should not have the method.

The method can exist, but it can be private. So this is really not something that needs to be debated now. As I said, the main strong possibility for reuse is with arrays of all of the same time. Maybe some day we'll need per-index type management, but I think that would just be a third method when we need it; I question if we'll ever need it.

There are no nullability checks in our builds, and nullability is known to be wrong many places and overall normally missing. Nullability annotations are a non-required annotation that just acts as a part of the javadoc, essentially. Nullability is good and fine, but it doesn't help anyone catch bugs any more than documentation.

It's also not clear to me why we need so much exception handling in `getNumberAsDoubleFromObj()`.

Renaming stuff to include `*FromObj` feels more like bikeshedding to me, or fixing a concern that the current callers don't suffer from. Some of those changes may be fine, but not as part of this change. That would be a separate refactor to the callers that should not change existing behaviors.

Reply:
> I don't understand what getFromObj() is providing, as obj.get(key) return null for non-present keys. This isn't Python. Note, if we had such a method it should throw an exception instead of returning null. The current methods consider null to be just another type and so fall-through to catch-all exceptions.

I think you missed the assert there. `getFromObj` does exactly what happens right now, it's a carbon copy of the first 5 lines of nearly every `get*` method in this class. So at least it's DRY.

Reply:
> The method can exist, but it can be private.
Yes, I'm totally ok with making this private. I was considering it, but then realized this is an internal class.

> It's also not clear to me why we need so much exception handling in getNumberAsDoubleFromObj().
Again, this just is exactly what we have right now, just wording in messages in the different order.

> Some of those changes may be fine, but not as part of this change
Yes, I started my comment with "This might be out of scope". However, unit-testing these private methods will be much more readable.

> Renaming stuff to include *FromObj feels more like bikeshedding to me
With private methods, we can keep existing name of the public methods as they are.

Reply:
I see what you did now. And the semantics in the existing code are a bit different than I thought. They return null when not available, so your suggested code would change behavior and start throwing `IllegalArgumentException`. We'd need every caller of `getFromObj()` to check for a `null` return.

```
Object o = getFromObj(obj, key);
if (o == null)
  return null;
```

Which is equivalent to (which is equivalent to the current code):
```
assert key != null;
Object o = obj.get(key);
if (o == null)
  return null;
```

Not much of a difference.

`assert key != null;` itself isn't really too interesting. Key should always be a hard-coded string, so isn't important for a nice error message or even (apparently) checking in production. If we cared more about it, it should be a checkNotNull.

There can be changes here, but it seems beside the purpose of this PR and not all that earth-shattering. Honestly, I'm surprised this has become a discussion point instead of something like implementing `getNumberAsInteger()` via `getNumberAsLong()` (or shared helper).

Reply:
```java
Object o = getFromObj(obj, key);
if (o == null)
  return null;
```

Ouch. I accidentally deleted this case from my example, when moved `getFromObj` into `try` at the very last moment. You're right, with the null check I missed it doesn't make a difference.

> not all that earth-shattering
It wasn't supposed to be :) This was a nit "hey, here we can break things up into smaller pieces, and don't do two things at a time".

Reply:
> Honestly, I'm surprised this has become a discussion point instead of something like implementing getNumberAsInteger() via getNumberAsLong() (or shared helper).

 To clarify: you mean this as an argument for the private methods that we discussed above?

Reply:
No, those private methods were for something different. I'm talking about how `getNumberAsInteger()` could be implemented as:

```java
Long l = getNumberAsLong(obj, key);
if (l == null)
  return null;
int i = l.intValue();
if ((long) i != l.longValue())
  throw ...;
return i;
```

Reply:
Note that longs this would produce different error messages, f.e: "value '%s' for key '%s' is not a **long** integer"

Reply:
Yeah, it was more of a sketch. And you have to figure out if the cost of merging is worth the benefit. I'm fine with it both ways (I think there are some ways that may not be too ugly to do it). Honestly, I'd be tempted to do it mostly because s/long/int/ is so error prone when comparing that logic.

@sergiitk and I talked. We're actually more on the same page than it seemed. The root of this is chain was really spawned by the tests, and I agree the tests exhibit some code smells that are a bit hard to track down the source for and the suggestions would improve things (for the tests).

The end result is probably not to change the API now. But we were also marveling at how many different issues something this simple brought up. Things could be improved with just changes to the tests, but we're probably not going to be overly critical because some of this should have been noticed with tests (that don't exist) introduced when JsonUtil was introduced. But we're also able to see "how we ended up here."

We should continue thinking about improvements to this area, but won't see changes right now.

Reply:
I agree there is space for improvement, but that's not something we should do in this PR.